algebird-core/src/main/scala/com/twitter/algebird/Preparer.scala

```diff
@@ -0,0 +1,235 @@
+  }
+
+  def min(implicit ord: Ordering[T]) = aggregate(Aggregator.min)
+  def minBy[U: Ordering](fn: T => U) = {
+    implicit val ordT = Ordering.by(fn)
+    aggregate(Aggregator.min[T])
+  }
+
+  def sum(implicit sg: Semigroup[T]) = aggregate(Aggregator.fromSemigroup(sg))
+  def reduce(fn: (T, T) => T) = aggregate(Aggregator.fromReduce(fn))
+}
+
+object MapPreparer {
+  /**
+   * Create a concrete MapPreparer.
+   */
+  def apply[A, T](fn: A => T) = new MapPreparer[A, T] { val prepareFn = fn }
+
+  /**
+   * This is purely an optimization for the case of mapping by identity.
+   * It overrides the key methods to not actually use the identity function.
+   */
+  def identity[A] = new MapPreparer[A, A] {
+    val prepareFn = (a: A) => a
+    override def map[U](fn: A => U) = MapPreparer(fn)
+    override def flatMap[U](fn: A => TraversableOnce[U]) = FlatMapPreparer(fn)
+    override def monoidAggregate[B, C](aggregator: MonoidAggregator[A, B, C]) = aggregator
+    override def aggregate[B, C](aggregator: Aggregator[A, B, C]) = aggregator
+  }
+}
+
+/**
+ * A Preparer that has had one or more flatMap operations applied.
+ * It can only accept MonoidAggregators.
+ */
+trait FlatMapPreparer[A, T] extends Preparer[A, T] {
+
+  def prepareFn: A => TraversableOnce[T]
+
+  def map[U](fn: T => U): FlatMapPreparer[A, U] =
+    FlatMapPreparer { a: A => prepareFn(a).map(fn) }
+
+  def flatMap[U](fn: T => TraversableOnce[U]) =
+    FlatMapPreparer { a: A => prepareFn(a).flatMap(fn) }
+
+  def monoidAggregate[B, C](aggregator: MonoidAggregator[T, B, C]): MonoidAggregator[A, B, C] =
+    aggregator.sumBefore.composePrepare(prepareFn)
+
+  /**
+   * alias of monoidAggregate for convenience
+   * unlike MapPreparer's aggregate, can only take MonoidAggregator
+   */
+  def aggregate[B, C](aggregator: MonoidAggregator[T, B, C]) = monoidAggregate(aggregator)
+
+  /**
+   * Like monoidAggregate, but using an implicit Monoid to construct the Aggregator
+   */
+  def sum(implicit monoid: Monoid[T]) = monoidAggregate(Aggregator.fromMonoid(monoid))
+
+  /**
+   * Split the processing into two parallel aggregations.
+   * You provide a function which produces two different aggregators from this preparer,
+   * and it will return a single aggregator which does both aggregations in parallel.
+   * (See also Aggregator's join method.)
+   *
+   * We really need to generate N versions of this for 3-way, 4-way etc splits.
+   */
+
+  def split[B1, B2, C1, C2](fn: FlatMapPreparer[TraversableOnce[T], T] => (MonoidAggregator[TraversableOnce[T], B1, C1], MonoidAggregator[TraversableOnce[T], B2, C2])): Aggregator[A, (B1, B2), (C1, C2)] = {
+    val (a1, a2) = fn(FlatMapPreparer.identity[T])
+    a1.join(a2).composePrepare(prepareFn)
+  }
+
+  /**
+   * transform a given Aggregator into a MonoidAggregator by lifting the reduce and present stages
+   * into Option space
+   */
+  def lift[B, C](aggregator: Aggregator[T, B, C]): MonoidAggregator[A, Option[B], Option[C]] =
```

Comment:
They could. Originally I had them without the 'Option' suffix (ie, `head`, not `headOption`) and so they would have conflicted with the semigroup versions. But I think it's better to be explicit.